design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,300 @@
+
+    Note: this does not cause the deletion of the OB.  Since the actual bucket is not deleted, it is reasonable that it should still tracked in Kubernetes.
+
+---
+
+## Looking Forward
+
+- Resource Quotas cannot yet be defined for CRDs.  This limits admin control over how many buckets can be created per object store.  A operator could be made configurable (via ConfigMap) to cap the number of buckets per object store.  A PR exists for enabling quotas on CRDs (https://github.com/kubernetes/kubernetes/pull/72384).  If it is merged into Kubernetes, a ResourceQuota example should be defined for OBCs.
+
+- Custom Bucket and Object policies are not defined for OBCs.  Currently all user keys will have Object PUT/GET/DELETE and object policy setting permissions.  It would be useful to allow users to link secondary keys with a subset of these permissions to buckets.
+
+- A generic S3 user API is not defined in this proposal but should be given serious consideration.  A user API should allow cluster users to generate new S3 users, tenants, and access keys in an object store.  This API could be used as method of manually creating object store users or incorporated into the bucket provisioning automation to generate the user's access key.
+
+---
+
+## API Specifications
+
+### OBC Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucketclaims.rook.io
+spec:
+  group: rook.io
+  version: v1
+  scope: namespaced
+  names:
+      kind: ObjectBucketClaim
+      listKind: ObjectBucketClaimList
+      singular: objectbucketclaim
+      plural: objectbucketclaims
+```
+
+
+### OBC Custom Resource (User Defined)
+
+```yaml
+apiVersion: rook.io/v1
+kind: ObjectBucketClaim
+metadata:
+  name: MY-BUCKET-1 [1]
+  namespace: USER-NAMESPACE [2]
+spec:
+  storageClassName: SOME-OBJECT-STORE-STORAGE-CLASS [3]
+  tenant: MY-TENANT [4]
+```
+
+1. `name` of the ObjectBucketClaim. Composes part of the bucket name and ConfigMap name
+1. `namespace` of the ObjectBucketClaim. Composes part of the bucket name. Determines the namespace of the ConfigMap
+1. `storageClassName` is used to target the desired Object Store.  Used by the operator to get the Object Store service URL.
+1. `tenant` allows users to define a tenant in an object store in order to namespace their buckets and access keys. 
+
+### OBC Custom Resource (Operator Updated)
+
+```yaml
+apiVersion: rook.io/v1
+kind: ObjectBucketClaim
+metadata:
+  name: MY-BUCKET-1
+  namespace: USER-NAMESPACE
+  ownerReferences: [1]
+  - name: CEPH-CLUSTER
+    ...
+spec:
+  storageClassName: SOME-OBJECT-STORE-STORAGE-CLASS
+status:
+  phase: {"pending", "bound", "error"}  [2]
+  objectBucketRef: objectReference{}  [3]
+  configMapRef: objectReference{}  [4]
+  secretRef: objectReference{}  [5]
+```
+
+1. `ownerReference` sets the ObjectBucketClaim as a child of the Ceph Cluster.  Cluster deletion causes the deletion of the OBC 
```

Comment:
If the user declared the OBC, would we really want the operator to update it with the ownerReference to be the cluster? If the user created it, seems like they should need to delete it as well. Rook's approach has been to add the ownerReference to resources it **generates**, rather than adding it to resources declared by the user.

Reply:
Good catch!  You're right, ownerRefs don't belong on user generated objects.  I'll remove this.